cuqipy_fenics/geometry.py

```diff
@@ -398,3 +403,222 @@ def _build_basis(self):
         # Normalize the eigenvectors if required
         if self.normalize:
             self._eig_vec /= np.linalg.norm( self._eig_vec )
+
+
+# Helper user expression to define the step expansion basis
+class _StepExpression(dl.UserExpression):
+    def __init__(self, degree, x_lim, y_lim=None):
+        self.x_lim = x_lim
+        self.y_lim = y_lim
+        super().__init__(degree=degree)
+
+    def eval(self, value, x):
+        if (
+            x[0] > self.x_lim[0] + dl.DOLFIN_EPS
+            and x[0] < self.x_lim[1] - dl.DOLFIN_EPS
+            and (self.y_lim is None or x[1] > self.y_lim[0] + dl.DOLFIN_EPS)
+            and (self.y_lim is None or x[1] < self.y_lim[1] - dl.DOLFIN_EPS)
+        ):
+            value[0] = 1
+        else:
+            value[0] = 0
+
+
+class StepExpansion(_WrappedGeometry):
+    """A geometry class that parameterizes finite element functions on a 1D or
+    2D domain with piecewise constant functions. In the 1D case, the domain is
```

Comment:
Good point. It works for example with circular mesh, I did not try other meshes. Note: I needed also to update the equality checks a bit in  `_StepExpression` class because some elements seemed not to belong to any basis before the update I did.
```python
# %%
import pytest
import numpy as np
from cuqipy_fenics.geometry import StepExpansion, FEniCSContinuous
from cuqi.distribution import Gaussian
import dolfin as dl
import  matplotlib.pyplot as plt
from cuqi.utilities import approx_gradient


num_steps_x = 4
num_steps_y = 4
if True:
    meshlevel = 100
    degree = 1
    dim = 2
    mesh = dl.UnitDiscMesh.create(dl.MPI.comm_world, meshlevel, degree, dim)
    # create circular mesh
    #mesh = dl.UnitDiscMesh(32)
    #mesh = dl.UnitSquareMesh(32,32) if num_steps_y is not None else dl.UnitIntervalMesh(32)
    V = dl.FunctionSpace(mesh, 'DG', 0)
    G_FEM = FEniCSContinuous(V, labels=['$\\xi_1$', '$\\xi_2$'])
    G_step = StepExpansion(G_FEM,
                                num_steps_x=num_steps_x, num_steps_y=num_steps_y)

    
    x = (
        np.arange(num_steps_x * num_steps_y)
        if num_steps_y is not None
        else np.arange(num_steps_x)
    )
    funvals = G_step.par2fun(x)
    im = dl.plot(funvals, title="StepExpansion")
    plt.colorbar(im)
```

![image](https://github.com/user-attachments/assets/003268e9-2dce-431a-8638-0ad0601e78ca)